In [10]:
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow.keras.layers import *
from tensorflow.keras import *
from tensorflow.keras.optimizers import Adam

print("TensorFlow version:", tf.__version__)
import ctypes

hllDll = ctypes.WinDLL("C:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v11.5\\bin\\cudart64_110.dll")
print(tf.config.list_physical_devices('GPU'))


TensorFlow version: 2.7.0
[]


In [7]:
dataset = pd.read_csv("./../../datasets/All.csv")

X = pd.DataFrame()
for col in ["Hour", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday", "SchoolHoliday", "Janurary", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December", "Cars-Total", "diesel", "wind_meters_per_second"]:
    X[col] = dataset[col].astype(float)

X["Cars-Total"] = X["Cars-Total"] / 1000
X["diesel"] = X["diesel"] / 1
X["wind_meters_per_second"] = X["wind_meters_per_second"] / 10

Y = pd.DataFrame()
for col in ["NO2-ug-per-m3"]:
    Y[col] = dataset[col].astype(float)

print(X.head())
print(Y.head())


   Hour  Monday  Tuesday  Wednesday  Thursday  Friday  Saturday  Sunday  \
0   4.0     0.0      1.0        0.0       0.0     0.0       0.0     0.0   
1   5.0     0.0      1.0        0.0       0.0     0.0       0.0     0.0   
2   6.0     0.0      1.0        0.0       0.0     0.0       0.0     0.0   
3   7.0     0.0      1.0        0.0       0.0     0.0       0.0     0.0   
4   8.0     0.0      1.0        0.0       0.0     0.0       0.0     0.0   

   SchoolHoliday  Janurary  ...  June  July  August  September  October  \
0            1.0       0.0  ...   0.0   0.0     0.0        0.0      0.0   
1            1.0       0.0  ...   0.0   0.0     0.0        0.0      0.0   
2            1.0       0.0  ...   0.0   0.0     0.0        0.0      0.0   
3            1.0       0.0  ...   0.0   0.0     0.0        0.0      0.0   
4            1.0       0.0  ...   0.0   0.0     0.0        0.0      0.0   

   November  December  Cars-Total    diesel  wind_meters_per_second  
0       0.0       0.0       

In [12]:
import matplotlib.pyplot as plt
import scipy.stats

hyper = {
    "input_dim": X.shape[1],
    "output_dim": Y.shape[1],
    "batch_size": 10,
    "epochs": 10,
    "learning_rate": 0.001,
    "validation_split": 0.2,
}

model = model = tf.keras.Sequential([
    Dense(hyper["input_dim"], activation='relu'),
    Dense(100, activation='sigmoid'),
    Dense(30, activation='sigmoid'),
    Dense(10, activation='sigmoid'),
    Dense(hyper["output_dim"], activation='sigmoid')
])

model.compile(
    optimizer=Adam(learning_rate=hyper["learning_rate"]),
    loss='mse',
    metrics=['mean_absolute_error'])

model.fit(
    X.to_numpy(),
    Y.to_numpy(),
    batch_size=hyper["batch_size"],
    epochs=hyper["epochs"],
    validation_split=hyper["validation_split"],
)

# plot guesses
predictions = model.predict(X.to_numpy())
predictions = np.abs(result - Y.to_numpy()) / Y.to_numpy()
for guesses in result[0, :]:
    guesses = np.sort(result[:, 0])
    std = str(round(x[int(len(x) * 0.682)]*100, 1),)
    plt.hist(x[x < 1], bins=100)
    plt.title("68% of predictions are less than "+std+"% off")
    plt.xlabel("Deviation from the label")
    plt.show()




Epoch 1/10
522/522 [==============================] - 2s 2ms/step - loss: 2829.7825 - mean_absolute_error: 47.2374 - val_loss: 2371.0398 - val_mean_absolute_error: 43.1303
Epoch 2/10
522/522 [==============================] - 1s 2ms/step - loss: 2824.8767 - mean_absolute_error: 47.1848 - val_loss: 2370.2075 - val_mean_absolute_error: 43.1205
Epoch 3/10
522/522 [==============================] - 1s 2ms/step - loss: 2824.3635 - mean_absolute_error: 47.1794 - val_loss: 2369.9514 - val_mean_absolute_error: 43.1177
Epoch 4/10
522/522 [==============================] - 1s 2ms/step - loss: 2824.1760 - mean_absolute_error: 47.1773 - val_loss: 2369.8384 - val_mean_absolute_error: 43.1162
Epoch 5/10
522/522 [==============================] - 1s 2ms/step - loss: 2824.0820 - mean_absolute_error: 47.1765 - val_loss: 2369.7776 - val_mean_absolute_error: 43.1155
Epoch 6/10
522/522 [==============================] - 1s 2ms/step - loss: 2824.0371 - mean_absolute_error: 47.1761 - val_loss: 2369.7424 - v

NameError: name 'result' is not defined

# Results

## Hour -> NO2
|Layers|Learning Rate|Epochs|Batch|Split|Sigma 1|
|---|---|---|---|---|---|
||||||

# Hour + Weekday -> NO2
|Layers|Learning Rate|Epochs|Batch|Split|Sigma 1|
|---|---|---|---|---|---|
||||||

# Hour + Weekday + Month + SchoolHoliday -> NO2
|Layers                                     |Learning Rate  |Epochs |Batch  |Split      |Sigma 1  | Loss  |
|---                                        |---            |---    |---    |---        |---        |---    |
| X relu; 50 relu ; 10 relu ; Y relu        |0.0001         |1000   |100    |0.2        |51%  | 430   |
| X relu; 50 relu ; 10 relu ; Y relu        |0.0005         |500    |100    |0.2        |51%  ||
| X relu; 50 relu ; 5  relu ; Y relu        |0.0005         |1000   |100    |0.2        |51%  | 415   |
| X relu; 50 relu ; 5  relu ; Y relu        |0.001          |1000   |100    |0.2        |51%  | 420   |
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu |0.0001      |100    |10     |0.2        |53%  | 423   |
| X relu; 50 relu ; 10 relu ; Y relu        |0.0005         |1000   |350    |0.2        |54%  | 420   |
| X relu; 50 relu ; 10 relu ; Y relu        |0.0003         |1000   |50     |0.2        |55%  | 420   |
| X relu; 50 sigm ; 10 sigm ; Y relu        |0.0005         |500    |2      |0.2        |56%  ||
| X relu; 30 relu ; 30 relu ; Y relu        |0.01           |20     |2      |0.2        |56%  ||
| X relu; 50 relu ; Y relu                  |0.0003         |500    |100    |0.2        |57%  ||
| X relu; 30 relu ; Y relu                  |0.0003         |500    |100    |0.2        |58%  ||
| X relu; 30 relu ; Y relu                  |0.001          |200    |100    |0.2        |58%  ||
| X relu; 30 relu ; 30 relu ; Y relu        |0.03           |20     |2      |0.2        |60%  ||
| X relu; 30 relu ; 30 relu ; Y relu        |0.001          |20     |2      |0.2        |61%  ||
| X relu; 30 relu ; 30 relu ; Y relu        |0.001          |20     |10     |0.2        |60%  ||
| X relu; 30 sigm ; 30 sigm ; Y relu        |0.01           |20     |10     |0.2        |60%  ||

# Hour + Weekday + Month + SchoolHoliday + Cars + WindSpeed + Diesel -> NO2
|Layers                                             |Learning Rate  |Epochs |Batch  |Split      |Sigma 1  | Loss  | Comment   |
|---                                                |---            |---    |---    |---        |---        |---    |---        |
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu      |0.001          |100    |10     |0.2        |36%  | 301   |Cras/1000 Wind/10 Diesel/2|
| X relu; 100 relu ; 50 relu ; 20 relu; Y relu      |0.001          |100    |10     |0.2        |36%  | 301   |Cras/1000 Wind/10 Diesel/2|
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu      |0.001          |100    |10     |0.2        |36%  | 304   |Cras/2000 Wind/20 Diesel/2|
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu      |0.0007         |100    |10     |0.2        |36%  | 301   |Cras/1000 Wind/10 Diesel/2|
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu      |0.003          |100    |10     |0.2        |37%  | 313   |Cras/1000 Wind/10 Diesel/2|
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu      |0.0003         |200    |20     |0.2        |37%  | 295   |Cras/1000 Wind/10 Diesel/2|
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu      |0.0005         |100    |10     |0.2        |40%  | 295   |Cras/1000 Wind/10 Diesel/2|
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu      |0.003          |200    |20     |0.2        |40%  | 337   |Cras/1000 Wind/10 Diesel/2|
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu      |0.001          |100    |10     |0.2        |40%  | 353   ||
| X relu; 100 relu ; 30 relu ; 10 relu; Y relu      |0.0001         |1000   |100    |0.2        |41%  | 348   ||
| X relu; 20  relu ; 10 relu ; 5 relu; Y relu       |0.0002         |1000   |100    |0.2        |41%  | 347   ||
| X relu; 50 relu ; 30 relu ; 10 relu; Y relu       |0.0001         |100    |5      |0.2        |41%  | 362   |Cras/1000 Wind/10 Diesel/2|
| X relu; 100 relu ; 10 relu ; Y relu               |0.0001         |1000   |100    |0.2        |41%  | 348   ||
| X relu; 50  relu ; 10 relu ; Y relu               |0.0001         |1000   |100    |0.2        |41%  | 352   ||
| X relu; 20  relu ; 10 relu ; 5 relu; Y relu       |0.001          |1000   |100    |0.2        |43%  | 336   ||
| X relu; 20  relu ; 10 relu ; 5 relu; Y relu       |0.0001         |1000   |100    |0.2        |44%  | 354   ||
| X relu; 50 relu ; 30 relu ; 10 relu; Y relu       |0.0001         |100    |20     |0.2        |46%  | 362   |Cras/1000 Wind/10 Diesel/2|
| X relu; 100 sigm ; 30 sigm ; 10 sigm; Y relu      |10.0           |100    |10     |0.2        |56%  | 766   |Cras/1000 Wind/10 Diesel/2|
